src/quark-core/src/QuarkScript.sol

```diff
@@ -58,16 +68,23 @@ abstract contract QuarkScript {
 
     function allowCallback() internal {
         QuarkWallet self = QuarkWallet(payable(address(this)));
-        self.stateManager().write(self.CALLBACK_KEY(), bytes32(uint256(uint160(self.stateManager().getActiveScript()))));
+        // TODO: Can save gas by just having the constant in QuarkScript
+        bytes32 callbackSlot = self.CALLBACK_SLOT();
+        bytes32 activeScriptSlot = self.ACTIVE_SCRIPT_SLOT();
+        assembly {
+            // TODO: Move to TLOAD/TSTORE after updating Solidity version to >=0.8.24
+            let activeScript := sload(activeScriptSlot)
+            sstore(callbackSlot, activeScript)
+        }
     }
 
     function clearCallback() internal {
         QuarkWallet self = QuarkWallet(payable(address(this)));
-        self.stateManager().write(self.CALLBACK_KEY(), bytes32(0));
-    }
-
-    function allowReplay() internal {
-        return QuarkWallet(payable(address(this))).stateManager().clearNonce();
+        bytes32 callbackSlot = self.CALLBACK_SLOT();
+        assembly {
+            // TODO: Move to TSTORE after updating Solidity version to >=0.8.24
```

Comment:
TSTORE should be double useful here, since it would never acccidentally be ON from a previous script, right?

Reply:
I'm curious about that, since each TSTORE is set from within a sub-transaction, right? I would expect (but don't know) that a TSTORE dies when it's owning sub-trx ends, as opposed to when the over-arching transaction ends, but maybe not?

Reply:
It actually behaves pretty much identical to SSTORE, so it persists for the entire transaction, not sub-transaction. This is from the EIP:

> Smart contract developers should understand the lifetime of transient storage variables before use. Because transient storage is automatically cleared at the end of the transaction, smart contract developers may be tempted to avoid clearing slots as part of a call in order to save gas. However, this could prevent further interactions with the contract in the same transaction (e.g. in the case of re-entrancy locks) or cause other bugs, so smart contract developers should be careful to only leave transient storage slots with nonzero values when those slots are intended to be used by future calls within the same transaction. Otherwise, these opcodes behave exactly the same as SSTORE and SLOAD, so all the usual security considerations apply especially in regard to reentrancy risk.